processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java

```diff
@@ -893,49 +939,150 @@ private static int compareDimsInRowsWithAggs(
-    private int[] sortableIds = null;
+    // dictionary id -> rank of the sorted dictionary
+    // This is initialized in the constructor and bufferComparator() with static dictionary and dynamic dictionary,
+    // respectively.
+    private int[] rankOfDictionaryIds = null;
 
     RowBasedKeySerde(
         final boolean includeTimestamp,
         final boolean sortByDimsFirst,
         final List<DimensionSpec> dimensions,
         final long maxDictionarySize,
         final DefaultLimitSpec limitSpec,
-        final List<ValueType> valueTypes
+        final List<ValueType> valueTypes,
+        @Nullable final List<String> dictionary
     )
     {
       this.includeTimestamp = includeTimestamp;
       this.sortByDimsFirst = sortByDimsFirst;
       this.dimensions = dimensions;
       this.dimCount = dimensions.size();
-      this.maxDictionarySize = maxDictionarySize;
       this.valueTypes = valueTypes;
       this.limitSpec = limitSpec;
-      this.serdeHelpers = makeSerdeHelpers();
+      this.enableRuntimeDictionaryGeneration = dictionary == null;
+      this.dictionary = enableRuntimeDictionaryGeneration ? new ArrayList<>(DICTIONARY_INITIAL_CAPACITY) : dictionary;
+      this.reverseDictionary = enableRuntimeDictionaryGeneration ?
+                               new Object2IntOpenHashMap<>(DICTIONARY_INITIAL_CAPACITY) :
+                               new Object2IntOpenHashMap<>(dictionary.size());
+      this.reverseDictionary.defaultReturnValue(UNKNOWN_DICTIONARY_ID);
+      this.maxDictionarySize = maxDictionarySize;
+      this.serdeHelpers = makeSerdeHelpers(limitSpec != null, enableRuntimeDictionaryGeneration);
+      this.serdeHelperComparators = new BufferComparator[serdeHelpers.length];
+      Arrays.setAll(serdeHelperComparators, i -> serdeHelpers[i].getBufferComparator());
       this.keySize = (includeTimestamp ? Longs.BYTES : 0) + getTotalKeySize();
       this.keyBuffer = ByteBuffer.allocate(keySize);
+
+      if (!enableRuntimeDictionaryGeneration) {
+        final long initialDictionarySize = dictionary.stream()
+                                                     .mapToLong(RowBasedGrouperHelper::estimateStringKeySize)
+                                                     .sum();
+        Preconditions.checkState(
+            maxDictionarySize >= initialDictionarySize,
+            "Dictionary size[%s] exceeds threshold[%s]",
+            initialDictionarySize,
+            maxDictionarySize
+        );
+
+        for (int i = 0; i < dictionary.size(); i++) {
+          reverseDictionary.put(dictionary.get(i), i);
+        }
+
+        initializeRankOfDictionaryIds();
+      }
+    }
+
+    private void initializeRankOfDictionaryIds()
+    {
+      final int dictionarySize = dictionary.size();
+      rankOfDictionaryIds = IntStream.range(0, dictionarySize).toArray();
+      IntArrays.quickSort(
+          rankOfDictionaryIds,
+          new IntComparator()
```

Comment:
I think version changing should be done in a separate issue. I'll raise a new issue for this.

Reply:
I don't see why it should be a separate issue and PR and more overhead, when it's super simple to update it here, and it would allow not to commit some ugly code.

Reply:
I don't think changing version is simple. It implies investigating what are the expected effects and testing everything works well beyond changing version numbers. It may be ok in this case because fastutil is a simple library providing fast collections, but I don't think this change is in a scope of this pr. 
Another reason is it's more simple to revert changes if any problems are found in the changed version.

Reply:
It's in the scope of this PR to not commit some ugly code. If problem is found in the library the bug should be fixed and version updated forward, not back.

Reply:
> It's in the scope of this PR to not commit some ugly code. 

I think the author should be the one who decides the scope of the pr, not the reviewer. Reviewers can't impose their opinions unless there is a significant problem. Otherwise, author's opinion should be respected.

Ugly code should be avoided, but is one of the most common tiny problems in development. It was added even in your patch (https://github.com/druid-io/druid/pull/4710#discussion_r135382922).
I didn't say I won't do that, but will do in a separate issue which means it will be fixed eventually. I don't understand why you stick to do in this pr.

> If problem is found in the library the bug should be fixed and version updated forward, not back.

Reverting version changes is a really common solution to workaround found problems until they are fixed.

Reply:
Actually, it's much better for finding the root cause. When we meet some problems but couldn't find the root cause easily, one of the ways to reduce the search coverage is going back through the git commit history. As each history contains more specific changes, it's easier find to the root cause.